File: src/main/java/net/sf/rails/game/state/WalletChange.java
package net.sf.rails.game.state;

public final class WalletChange<T extends Countable> extends Change {

    private final Wallet<T> wallet;
    private final T item;
    private final int amount;
    
    WalletChange(Wallet<T> wallet, T item, int amount) {
        this.wallet = wallet;
        this.item = item;
        this.amount = amount;
        super.init(wallet);
    }
    
    @Override 
    void execute() {
        wallet.change(item, amount);
    }

    @Override 
    void undo() {
        wallet.change(item, -amount);
    }

    @Override
    public 
    Wallet<T> getState() {
        return wallet;
    }
    
    @Override
    public String toString() {
        return "Change for " + wallet + ": " + amount + " of " + item ;
    }
}

File: dotnet/RailsLib.Net/Net/Game/State/WalletChange.cs
using System;


namespace GameLib.Net.Game.State
{
    public class WalletChange<T> : Change where T : ICountable
    {
        private readonly Wallet<T> wallet;
        private readonly T item;
        private readonly int amount;

        public WalletChange(Wallet<T> wallet, T item, int amount)
        {
            this.wallet = wallet;
            this.item = item;
            this.amount = amount;
            base.Init(wallet);
        }

        override public void Execute()
        {
            wallet.Change(item, amount);
        }


        override public void Undo()
        {
            wallet.Change(item, -amount);
        }

        public override GameState GameState
        {
            get
            {
                return wallet;
            }
        }

        override public string ToString()
        {
            return "Change for " + wallet + ": " + amount + " of " + item;
        }
    }
}
